File: sdks/elixir/lib/mongo_do/cursor.ex
defmodule MongoDo.Cursor do
  @moduledoc """
  Cursor for iterating over MongoDB query results.

  Implements the Enumerable protocol for lazy iteration.

  ## Examples

      {:ok, cursor} = Collection.find(users, %{active: true})

      cursor
      |> Enum.take(10)
      |> Enum.each(&process_user/1)
  """

  @type t :: %__MODULE__{
          id: term(),
          client: MongoDo.Client.t(),
          database: String.t(),
          collection: String.t(),
          batch: [map()],
          exhausted: boolean()
        }

  defstruct [:id, :client, :database, :collection, batch: [], exhausted: false]

  defimpl Enumerable do
    def count(_cursor), do: {:error, __MODULE__}
    def member?(_cursor, _element), do: {:error, __MODULE__}
    def slice(_cursor), do: {:error, __MODULE__}

    def reduce(_cursor, {:halt, acc}, _fun) do
      {:halted, acc}
    end

    def reduce(cursor, {:suspend, acc}, fun) do
      {:suspended, acc, &reduce(cursor, &1, fun)}
    end

    def reduce(%MongoDo.Cursor{exhausted: true, batch: []}, {:cont, acc}, _fun) do
      {:done, acc}
    end

    def reduce(%MongoDo.Cursor{batch: [doc | rest]} = cursor, {:cont, acc}, fun) do
      reduce(%{cursor | batch: rest}, fun.(doc, acc), fun)
    end

    def reduce(%MongoDo.Cursor{batch: [], exhausted: false} = cursor, {:cont, acc}, fun) do
      case MongoDo.Cursor.get_more(cursor) do
        {:ok, %{batch: [], exhausted: true} = cursor} ->
          {:done, acc}

        {:ok, cursor} ->
          reduce(cursor, {:cont, acc}, fun)

        {:error, _reason} ->
          {:done, acc}
      end
    end
  end

  @doc """
  Create a new cursor.
  """
  @spec new(map()) :: t()
  def new(opts) do
    struct(__MODULE__, opts)
  end

  @doc """
  Get the next batch of documents from the cursor.
  """
  @spec get_more(t()) :: {:ok, t()} | {:error, term()}
  def get_more(%__MODULE__{exhausted: true} = cursor) do
    {:ok, cursor}
  end

  def get_more(%__MODULE__{} = cursor) do
    case MongoDo.Client.execute(cursor.client, {:get_more, cursor.id, cursor.database, cursor.collection}) do
      {:ok, %{"cursor" => %{"id" => 0, "nextBatch" => batch}}} ->
        {:ok, %{cursor | batch: batch, exhausted: true}}

      {:ok, %{"cursor" => %{"id" => id, "nextBatch" => batch}}} ->
        {:ok, %{cursor | id: id, batch: batch}}

      {:ok, %{"nextBatch" => batch}} when batch == [] ->
        {:ok, %{cursor | batch: [], exhausted: true}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Close the cursor and release server resources.
  """
  @spec close(t()) :: :ok
  def close(%__MODULE__{id: nil}), do: :ok
  def close(%__MODULE__{exhausted: true}), do: :ok

  def close(%__MODULE__{} = cursor) do
    MongoDo.Client.execute(cursor.client, {:kill_cursors, [cursor.id]})
    :ok
  end

  @doc """
  Convert cursor to a list (fetches all documents).
  """
  @spec to_list(t()) :: [map()]
  def to_list(%__MODULE__{} = cursor) do
    Enum.to_list(cursor)
  end

  @doc """
  Check if cursor has more documents.
  """
  @spec has_next?(t()) :: boolean()
  def has_next?(%__MODULE__{batch: [_ | _]}), do: true
  def has_next?(%__MODULE__{exhausted: true}), do: false
  def has_next?(_cursor), do: true
end

File: sdks/clojure/src/mongo_do/cursor.clj
(ns mongo-do.cursor
  "Cursor implementation for lazy sequence iteration over MongoDB results.

   Provides lazy sequences for memory-efficient processing of large result sets.")

;; ============================================================================
;; Cursor State
;; ============================================================================

(defn- fetch-batch
  "Fetch a batch of documents from the server."
  [rpc db-name coll-name filter opts cursor-id batch-size]
  (let [query-opts (cond-> (or opts {})
                     cursor-id (assoc :cursor-id cursor-id)
                     batch-size (assoc :batch-size batch-size))
        result (-> rpc .-mongo (.find db-name coll-name filter query-opts))]
    (if (map? result)
      ;; Cursor-based result
      {:documents (get result :documents [])
       :cursor-id (get result :cursor-id)
       :has-more (get result :has-more false)}
      ;; Simple array result
      {:documents (if (sequential? result) (vec result) [])
       :cursor-id nil
       :has-more false})))

;; ============================================================================
;; Lazy Sequence Implementation
;; ============================================================================

(defn ->lazy-seq
  "Create a lazy sequence of documents from a MongoDB query.

   The sequence fetches documents in batches for efficiency.

   Args:
     rpc - RPC client
     db-name - Database name
     coll-name - Collection name
     filter - Query filter map
     opts - Query options (projection, sort, limit, etc.)

   Returns:
     Lazy sequence of document maps."
  [rpc db-name coll-name filter opts]
  (let [batch-size (get opts :batch-size 100)
        limit (get opts :limit)
        seen (atom 0)]
    (letfn [(fetch-next [cursor-id]
              (lazy-seq
                (when (or (nil? limit) (< @seen limit))
                  (let [{:keys [documents cursor-id has-more]}
                        (fetch-batch rpc db-name coll-name filter opts cursor-id batch-size)
                        remaining (if limit
                                    (min (count documents) (- limit @seen))
                                    (count documents))
                        docs (take remaining documents)]
                    (swap! seen + (count docs))
                    (if (seq docs)
                      (concat docs
                              (when (and has-more
                                         (or (nil? limit) (< @seen limit)))
                                (fetch-next cursor-id)))
                      nil)))))]
      (fetch-next nil))))

;; ============================================================================
;; Cursor Object (for explicit control)
;; ============================================================================

(defprotocol ICursor
  "Protocol for cursor operations."
  (-has-next? [this] "Check if there are more documents.")
  (-next [this] "Get the next document.")
  (-close! [this] "Close the cursor.")
  (-to-seq [this] "Convert to lazy sequence.")
  (-to-vec [this] "Fetch all remaining documents as a vector."))

(defrecord Cursor [rpc db-name coll-name filter opts
                   ^:volatile-mutable documents
                   ^:volatile-mutable cursor-id
                   ^:volatile-mutable has-more
                   ^:volatile-mutable closed]
  ICursor
  (-has-next? [this]
    (cond
      closed false
      (seq documents) true
      (not has-more) false
      :else (do
              ;; Fetch next batch
              (let [batch (fetch-batch rpc db-name coll-name filter opts cursor-id
                                       (get opts :batch-size 100))]
                (set! documents (:documents batch))
                (set! cursor-id (:cursor-id batch))
                (set! has-more (:has-more batch))
                (seq documents)))))

  (-next [this]
    (when (-has-next? this)
      (let [doc (first documents)]
        (set! documents (rest documents))
        doc)))

  (-close! [this]
    (set! closed true)
    (set! documents nil)
    nil)

  (-to-seq [this]
    (lazy-seq
      (when (-has-next? this)
        (cons (-next this) (-to-seq this)))))

  (-to-vec [this]
    (loop [result []]
      (if (-has-next? this)
        (recur (conj result (-next this)))
        result))))

(defn cursor
  "Create a cursor for iterating over query results.

   The cursor provides explicit control over iteration and cleanup.

   Args:
     rpc - RPC client
     db-name - Database name
     coll-name - Collection name
     filter - Query filter map
     opts - Query options

   Returns:
     Cursor instance.

   Example:
     (let [c (cursor rpc \"mydb\" \"users\" {:status \"active\"} nil)]
       (try
         (while (has-next? c)
           (println (next! c)))
         (finally
           (close! c))))"
  [rpc db-name coll-name filter opts]
  (->Cursor rpc db-name coll-name filter opts
            nil nil true false))

(defn has-next?
  "Check if the cursor has more documents."
  [cursor]
  (-has-next? cursor))

(defn next!
  "Get the next document from the cursor."
  [cursor]
  (-next cursor))

(defn close!
  "Close the cursor and release resources."
  [cursor]
  (-close! cursor))

(defn to-seq
  "Convert cursor to lazy sequence."
  [cursor]
  (-to-seq cursor))

(defn to-vec
  "Fetch all remaining documents as a vector."
  [cursor]
  (-to-vec cursor))

;; ============================================================================
;; with-cursor Macro
;; ============================================================================

(defmacro with-cursor
  "Execute body with a cursor, ensuring cleanup.

   Example:
     (with-cursor [c (cursor rpc db coll filter opts)]
       (doseq [doc (to-seq c)]
         (println doc)))"
  [[binding cursor-expr] & body]
  `(let [~binding ~cursor-expr]
     (try
       ~@body
       (finally
         (close! ~binding)))))
